stan/math/opencl/kernel_generator/unary_operation_cl.hpp

```diff
@@ -0,0 +1,170 @@
+  using base::var_name;
+
+  /**
+   * Constructor
+   * @param a argument expression
+   * @param fun function
+   */
+  unary_operation_cl(T&& a, const std::string& op)
+      : base(std::forward<T>(a)), op_(op) {}
+
+  /**
+   * generates kernel code for this expression.
+   * @param i row index variable name
+   * @param j column index variable name
+   * @param var_name_arg variable name of the nested expression
+   * @return part of kernel with code for this expression
+   */
+  inline kernel_parts generate(const std::string& i, const std::string& j,
+                               const std::string& var_name_arg) const {
+    kernel_parts res{};
+    res.body = type_str<Scalar>() + " " + var_name + " = " + op_ + var_name_arg
+               + ";\n";
+    return res;
+  }
+
+ protected:
+  std::string op_;
+};
+
+/**
+ * Represents a logical negation in kernel generator expressions.
+ * @tparam Derived derived type
+ * @tparam T type of argument
+ */
+template <typename T>
+class logical_negation_
+    : public unary_operation_cl<logical_negation_<T>, T, bool> {
+  static_assert(
+      std::is_integral<typename std::remove_reference_t<T>::Scalar>::value,
+      "logical_negation: argument must be expression with integral "
+      "or boolean return type!");
+  using base = unary_operation_cl<logical_negation_<T>, T, bool>;
+  using base::arguments_;
+
+ public:
+  /**
+   * Constructor
+   * @param a argument expression
+   */
+  explicit logical_negation_(T&& a) : base(std::forward<T>(a), "!") {}
+
+  /**
+   * Creates a deep copy of this expression.
+   * @return copy of \c *this
+   */
+  inline auto deep_copy() const {
+    auto&& arg_copy = this->template get_arg<0>().deep_copy();
+    return logical_negation_<std::remove_reference_t<decltype(arg_copy)>>{
+        std::move(arg_copy)};
+  }
+
+  /**
+   * View of a matrix that would be the result of evaluating this expression.
+   * @return view
+   */
+  inline matrix_cl_view view() const { return matrix_cl_view::Entire; }
+};
+
+/**
+ * Logical negation of a kernel generator expression.
+ *
+ * @tparam T type of the argument
+ * @param a argument expression
+ * @return logical negation of given expression
+ */
+template <typename T,
+          require_all_valid_expressions_and_none_scalar_t<T>* = nullptr>
+inline logical_negation_<as_operation_cl_t<T>> operator!(T&& a) {
+  return logical_negation_<as_operation_cl_t<T>>(
+      as_operation_cl(std::forward<T>(a)));
+}
+
+/**
+ * Represents an unary minus operation in kernel generator expressions.
+ * @tparam Derived derived type
+ * @tparam T type of argument
+ */
+template <typename T>
+class unary_minus_
+    : public unary_operation_cl<unary_minus_<T>, T,
+                                typename std::remove_reference_t<T>::Scalar> {
+  using base = unary_operation_cl<unary_minus_<T>, T,
+                                  typename std::remove_reference_t<T>::Scalar>;
+  using base::arguments_;
```

Comment:
Does this get used somewhere else?

Reply:
I was planning to do a refactor that would make derived classes use `operation_cl::arguments_` directely again instead of `get_argument` (possibly before this PR was merged). At the end I did not do it in this PR, but it is comming.